### ISSUE TITLE
Namespace qualification for symbol references

### ISSUE THREAD

**Issue by hvr** [haskell/haddock#667]
## The Problem

Right now, we distinguish only between 3 classes

 1. (by using `"` quotes) module names
 2. (by using `'` quotes) lower-cased symbols, i.e. ordinary term level bindings    
 3. (by using `'` quotes) upper-cased symbols: class names, type names, data constructors, and others

The problem however is that 3. conflates more than one namespace (regardless of H2010 forbidding an identifier to refer to a class and constructor within the same scope); specifically you can still end up in a module having separate entities have the same identifier; e.g. in the following example,

```hs
module Bar where

class Foo t where

data Bar = Foo

data Doo = Bar
```

the Haddock identifier references

 - `'Bar.Bar'` and
 - `'Bar.Foo'`

are **not** uniquely determined.

Moreover, when referring to identifiers from packages which aren't in scope, Haddock needs to make a guess.

Finally, for rendering identifier references in haddock-marked-up  `.cabal` descriptions, hackage-server cannot perform any type-inference, so a way to decide whether to use a t: or a v: style href is needed.

UPDATE: The necessity for this disambiguation has been increased due to HiHaddock which moves the identifier resolving/lookup into the GHC compilation pipeline at which point references in Haddock docstrings need to be resolved by GHC in a well-defined matter. Blindly guessing the category of an identifier doesn't fit well into this new scheme any more; we need to start treating docstring references in a more principled way than the stringly-typed way Haddock did previously.

### Variant a

Introduce a way/syntax to qualify the namespace of upper-case identifiers

Bikesheddable suggestion (which follows the way the current link-anchor generation works) for qualified identifier references:

 - `'t:Foo'` (or `'t:Bar.Foo'`)  refers to `class Foo`
 - `'v:Foo'` (or `'v:Bar.Foo'`) refers to `Foo :: Bar`
 - `'t:Bar'` (or `'t:Bar.Bar'`) refers to `data Bar`
 - `'v:Bar'`(or `'v:Bar.Bar'`) refers to `Bar :: Doo`

Haddock would warn whenever it cannot uniquely disambiguate a reference, so that users know when it's needed to qualify a given reference by `v:` or `t:`.

### Variant b

Use  export/import-list-ish syntax to qualify value level terms for upper-case symbols (which otherwise by default (or in case of ambiguity) refer to non-value level symbols):

 - `'(Foo)'` or `'Bar(Foo)'` or `'Bar.(Foo)'` or `'Bar.Bar(Foo)'` refers to `Foo :: Bar`
 - `'(Bar)'` or `'Bar.(Bar)'` or `'Doo(Bar)'` or `'Bar.Doo(Bar)'` refers to `Bar :: Doo`

NOTE:
 - We allow to leave off the type name before the `()`s
 - We don't have a way to explicitly qualify non-value-level symbols; consequently we cannot warn about ambiguities as there's no actionable way to silence those warnings if we in fact want to refer to non-value symbols.

### Variant c

(mentioned in https://github.com/haskell/haddock/issues/667#issuecomment-400238387)

This is a more economic&ergonomic version of *Variant a* which is inspired by Python-style literal prefixes:

 - `t'Foo'` (or `t'Bar.Foo'`)  refers to `class Foo`
 - `v'Foo'` (or `v'Bar.Foo'`) refers to `Foo :: Bar`
 - `t'Bar'` (or `t'Bar.Bar'`) refers to `data Bar`
 - `v'Bar'`(or `v'Bar.Bar'`) refers to `Bar :: Doo`

This also avoids the awkwardness for referring to operators as mentioned in https://github.com/haskell/haddock/issues/667#issuecomment-399502298:

 - `v':='`
 - `v'+'`

This exploits that this markup only requires to prefix merely a single letter for disambiguating and is extensible to additional qualifications if the need should ever arise. Another benefit of this variant is  graceful degradation with  older Haddock parsers which will still recognize the symbol reference and resolve it according to the old heuristics into hyperlinks; neither *Variant a* nor *Variant b* have this property.

I don't think we can do any better than *Variant c*; hence why I believe this to be the most convenient markup possible in terms of editing/typing friendliness. It's also reasonably non-intrusive and doesn't steal syntax you'd otherwise use in English prose typesetting.

**Comment by hvr** [comment]
@alexbiehl fyi, I've added a "variant b"

**Comment by sjakobi** [comment]
So far I think variant __a__ looks nicer.

What about operators though? `'+'` could be written as `'v:+'`? Constructor operators would look a bit funny, e.g. `'v::='` for `':='`. Not sure whether that's a reason to use a different separator like `,` or `;` (which have the advantage of not being valid operator symbols).

**Comment by sjakobi** [comment]
I just talked to Alex and Herbert about this. We came up with a different syntax that resembles Python's unicode qualifiers (`u"foo"`):

`t'Foo'` or `v'+'`.

Herbert mentioned he would try to implement it in the parser.

**Comment by harpocrates** [comment]
@glguy has another suggestion: we could follow the lead of `ExplicitNamespaces` and use the `type` keyword. Basically:

  * `'Foo'` first tries to link to a value, then to a type if that fails (so _not_ the existing default behaviour), and then falls back onto monospaced `Foo`.
  * `'type Foo'` tries to link to a type, and then falls back onto monospaced `Foo`.

Another argument this idea has going for it: we could eventually fix `DEPRECATED` pragmas to have this convention too.

**Comment by theindigamer** [comment]
I don't have a preference for a particular syntax but emitting a warning on ambiguity sounds _very useful_ because I routinely forget small stuff like this.

**Comment by sjakobi** [comment]
> I don't have a preference for a particular syntax but emitting a warning on ambiguity sounds very useful because I routinely forget small stuff like this.

The upcoming haddock release for ghc-8.6 will already include such a warning.

**Comment by theindigamer** [comment]
I see. I was concerned about hvr's Variant b because he wrote that it won't be possible to warn about it. If warnings will stay, then that's great.

**Comment by sjakobi** [comment]
> I see. I was concerned about hvr's Variant b because he wrote that it won't be possible to warn about it. If warnings will stay, then that's great.

Ah ok. I missed the context of your comment.

**Comment by mgsloan** [comment]
I prefer variant a (`v:Foo` / `t:Foo`), because it matches haddock's anchor links.  It seems the most consistent - less things to learn.

**Comment by mgsloan** [comment]
In [a reddit comment](https://www.reddit.com/r/haskell/comments/90b7rh/bikeshedding_haddock_syntax_for_symbol_references/e2p8v7c/) I described how a markdown variant of haddock syntax would be very good.  In anticipation of that approach, I propose expanding this bikeshed to include variant C:

## Variant C

* `'Foo'(t)` (or `'Bar.Foo'(t)`) refers to `class Foo`
* `'Foo'(v)` (or `'Bar.Foo'(v)`) refers to `Foo :: Bar`
* `'Bar'(t)` (or `'Bar.Bar'(t)`) refers to `data Bar`
* `'Bar'(v)` (or `'Bar.Bar'(v)`) refers to `Bar :: Doo`

It is also worth considering accepting more explicit `'Foo'(type)` or `'Foo'(value)` as well.

I think I prefer how this looks compared to Variant A, but I also like that Variant A is consistent with the anchor syntax.  Perhaps an eventual transition to markdown based haddock is not the best way to go, but I think that would be a very positive direction for making haddock more usable and predictable.  I particularly like that my proposal drastically reduces the quoting you have to learn - all you have is backticks and everything is contextually inferred with explicit disambiguation when necessary.

**Comment by sjakobi** [comment]
@mgsloan: You can find the markdown discussion at https://github.com/haskell/haddock/pull/794. A small prototype was implemented in https://github.com/haskell/haddock/pull/729.

**Comment by mgsloan** [comment]
@sjakobi Oh fantastic, thanks for the links!  I knew of efforts years ago but I'm super glad to see this return.

**Comment by harpocrates** [comment]
Fixed in dd47029cb29c80b1ab4db520c9c2ce4dca37f833. In the end, we went with variant C ffrom [the initial bug report](https://github.com/haskell/haddock/issues/667#issue-249822659).
